File: Core/Models/TrainingLoadEntry.swift
import Foundation

public enum SessionType: String, Codable, CaseIterable, Sendable, Hashable {
    case technique, sparring, fitness, poomsae, mixed

    public var labelKey: String { "session_type.\(rawValue)" }

    public var systemIcon: String {
        switch self {
        case .technique: "figure.kickboxing"
        case .sparring: "figure.boxing"
        case .fitness: "figure.strengthtraining.traditional"
        case .poomsae: "figure.taichi"
        case .mixed: "circle.grid.2x2"
        }
    }
}

public struct TrainingLoadEntry: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var athleteID: EntityID
    /// Optional link to a `ClassSession` row when the entry corresponds to a
    /// scheduled session. Standalone entries (home practice, supplementary
    /// work) leave this nil.
    public var sessionID: EntityID?
    public var recordedAt: Date
    public var sessionType: SessionType
    public var durationMinutes: Int
    /// 1...10 — Borg-style perceived exertion.
    public var rpe: Int
    public var notes: String?

    public init(
        id: EntityID = UUID(),
        athleteID: EntityID,
        sessionID: EntityID? = nil,
        recordedAt: Date,
        sessionType: SessionType,
        durationMinutes: Int,
        rpe: Int,
        notes: String? = nil
    ) {
        self.id = id
        self.athleteID = athleteID
        self.sessionID = sessionID
        self.recordedAt = recordedAt
        self.sessionType = sessionType
        self.durationMinutes = max(0, durationMinutes)
        self.rpe = max(1, min(10, rpe))
        self.notes = notes
    }

    /// Foster's session-RPE: duration (min) × RPE → arbitrary units (AU).
    public var sessionLoad: Double {
        Double(durationMinutes) * Double(rpe)
    }
}

/// Risk classification from the acute:chronic workload ratio (ACWR). The
/// commonly cited "sweet spot" is 0.8–1.3; >1.5 is the user-defined danger
/// zone. Below 0.8 indicates undertraining (deconditioning risk).
public enum LoadRisk: String, Sendable, Hashable {
    case undertrained, sweet, elevated, danger, unknown

    public var labelKey: String { "load_risk.\(rawValue)" }
}

public extension Array where Element == TrainingLoadEntry {

    /// Sum of session loads in the last `days` days (default 7), ending at `asOf`.
    func acuteLoad(asOf date: Date = Date(), days: Int = 7) -> Double {
        let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: date) ?? date
        return self.filter { $0.recordedAt > cutoff && $0.recordedAt <= date }
            .map(\.sessionLoad)
            .reduce(0, +)
    }

    /// Sum of session loads in the last `days` days (default 28).
    func chronicLoad(asOf date: Date = Date(), days: Int = 28) -> Double {
        let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: date) ?? date
        return self.filter { $0.recordedAt > cutoff && $0.recordedAt <= date }
            .map(\.sessionLoad)
            .reduce(0, +)
    }

    /// ACWR = (7-day weekly average) / (28-day weekly average)
    ///       = (7-day sum / 1) / (28-day sum / 4)
    /// Returns nil when chronic load is zero (insufficient history).
    func acwr(asOf date: Date = Date()) -> Double? {
        let acute = acuteLoad(asOf: date)
        let chronicWeekly = chronicLoad(asOf: date) / 4.0
        guard chronicWeekly > 0 else { return nil }
        return acute / chronicWeekly
    }

    func loadRisk(asOf date: Date = Date()) -> LoadRisk {
        guard let r = acwr(asOf: date) else { return .unknown }
        switch r {
        case ..<0.8: return .undertrained
        case 0.8..<1.3: return .sweet
        case 1.3..<1.5: return .elevated
        default: return .danger
        }
    }
}

File: flutter_app/lib/core/models/training_load_entry.dart
import 'entity_id.dart';

/// Port of Core/Models/TrainingLoadEntry.swift.

enum SessionType {
  technique,
  sparring,
  fitness,
  poomsae,
  mixed;

  String get labelKey => 'session_type.$name';

  String get systemIcon => switch (this) {
    SessionType.technique => 'figure.kickboxing',
    SessionType.sparring => 'figure.boxing',
    SessionType.fitness => 'figure.strengthtraining.traditional',
    SessionType.poomsae => 'figure.taichi',
    SessionType.mixed => 'circle.grid.2x2',
  };

  static SessionType fromJson(String raw) => SessionType.values.firstWhere(
    (e) => e.name == raw,
    orElse: () => SessionType.mixed,
  );
}

class TrainingLoadEntry {
  final EntityID id;
  final EntityID athleteId;

  /// Optional link to a ClassSession row when the entry corresponds to a
  /// scheduled session. Standalone entries (home practice, supplementary
  /// work) leave this null.
  final EntityID? sessionId;
  final DateTime recordedAt;
  final SessionType sessionType;
  final int durationMinutes;

  /// 1…10 — Borg-style perceived exertion.
  final int rpe;

  final String? notes;

  TrainingLoadEntry({
    EntityID? id,
    required this.athleteId,
    this.sessionId,
    required this.recordedAt,
    required this.sessionType,
    required int durationMinutes,
    required int rpe,
    this.notes,
  }) : id = id ?? newEntityId(),
       durationMinutes = durationMinutes < 0 ? 0 : durationMinutes,
       rpe = rpe.clamp(1, 10);

  /// Foster's session-RPE: duration (min) × RPE → arbitrary units (AU).
  double get sessionLoad => durationMinutes * rpe.toDouble();

  factory TrainingLoadEntry.fromJson(Map<String, dynamic> json) =>
      TrainingLoadEntry(
        id: json['id'] as String,
        athleteId: json['athleteID'] as String,
        sessionId: json['sessionID'] as String?,
        recordedAt: DateTime.parse(json['recordedAt'] as String),
        sessionType: SessionType.fromJson(json['sessionType'] as String),
        durationMinutes: json['durationMinutes'] as int,
        rpe: json['rpe'] as int,
        notes: json['notes'] as String?,
      );

  Map<String, dynamic> toJson() => {
    'id': id,
    'athleteID': athleteId,
    'sessionID': sessionId,
    'recordedAt': recordedAt.toIso8601String(),
    'sessionType': sessionType.name,
    'durationMinutes': durationMinutes,
    'rpe': rpe,
    'notes': notes,
  };
}

/// Risk classification from the acute:chronic workload ratio (ACWR).
/// The commonly cited "sweet spot" is 0.8–1.3; >1.5 is the danger zone.
/// Below 0.8 indicates undertraining (deconditioning risk).
///
/// NOTE: LoadRisk is NOT Codable in Swift — it is NOT given fromJson/toJson.
enum LoadRisk {
  undertrained,
  sweet,
  elevated,
  danger,
  unknown;

  String get labelKey => 'load_risk.$name';
}

extension TrainingLoadEntryListExt on List<TrainingLoadEntry> {
  /// Sum of session loads in the last [days] days ending at [asOf].
  double acuteLoad({DateTime? asOf, int days = 7}) {
    final date = asOf ?? DateTime.now();
    final cutoff = date.subtract(Duration(days: days));
    return where(
      (e) => e.recordedAt.isAfter(cutoff) && !e.recordedAt.isAfter(date),
    ).fold(0.0, (sum, e) => sum + e.sessionLoad);
  }

  /// Sum of session loads in the last [days] days (default 28).
  double chronicLoad({DateTime? asOf, int days = 28}) {
    final date = asOf ?? DateTime.now();
    final cutoff = date.subtract(Duration(days: days));
    return where(
      (e) => e.recordedAt.isAfter(cutoff) && !e.recordedAt.isAfter(date),
    ).fold(0.0, (sum, e) => sum + e.sessionLoad);
  }

  /// ACWR = (7-day weekly average) / (28-day weekly average).
  /// Returns null when chronic load is zero (insufficient history).
  double? acwr({DateTime? asOf}) {
    final date = asOf ?? DateTime.now();
    final acute = acuteLoad(asOf: date);
    final chronicWeekly = chronicLoad(asOf: date) / 4.0;
    if (chronicWeekly <= 0) return null;
    return acute / chronicWeekly;
  }

  LoadRisk loadRisk({DateTime? asOf}) {
    final r = acwr(asOf: asOf);
    if (r == null) return LoadRisk.unknown;
    if (r < 0.8) return LoadRisk.undertrained;
    if (r < 1.3) return LoadRisk.sweet;
    if (r < 1.5) return LoadRisk.elevated;
    return LoadRisk.danger;
  }
}
